src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +201,492 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
 
 
 /*
+ * Creates a router plan for INSERT ... SELECT queries which can consists of
+ * multiple tasks.
+ *
+ * The function never returns NULL, it errors out if cannot create the multi plan.
+ */
+static MultiPlan *
+CreateMultiTaskRouterPlan(Query *originalQuery, Query *query,
+						  RelationRestrictionContext *restrictionContext)
+{
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	int shardOffset = 0;
+	int shardCount = cacheEntry->shardIntervalArrayLength;
+	List *sqlTaskList = NIL;
+	uint32 taskIdIndex = 1;     /* 0 is reserved for invalid taskId */
+	Job *workerJob = NULL;
+	uint64 jobId = INVALID_JOB_ID;
+	MultiPlan *multiPlan = NULL;
+
+	/*
+	 * Error semantics for INSERT ... SELECT queries are different than regular
+	 * modify queries. Thus, handle separately.
+	 */
+	ErrorIfInsertSelectQueryNotSupported(originalQuery);
+
+	/*
+	 * Plan select query for each shard in the target table. Do so by
+	 * replacing the magic parameters added in multi_planner() with actual
+	 * current shard's boundary values. Then perform the normal shard
+	 * pruning.
+	 */
+	for (shardOffset = 0; shardOffset < shardCount; shardOffset++)
+	{
+		Query *subquery = ((RangeTblEntry *) list_nth(query->rtable, 1))->subquery;
+		Query *copiedOriginal = copyObject(originalQuery);
+		Query *originalSubquery = ((RangeTblEntry *) list_nth(copiedOriginal->rtable,
+															  1))->subquery;
+		RelationRestrictionContext *copiedRestrictionContext =
+			copyRelationRestrictionContext(restrictionContext);
+		ShardInterval *shardInterval =
+			cacheEntry->sortedShardIntervalArray[shardOffset];
+		uint64 shardId = shardInterval->shardId;
+		StringInfo queryString = makeStringInfo();
+		ListCell *restrictionCell = NULL;
+		Query *routerQuery = NULL;
+		Task *sqlTask = NULL;
+		List *selectPlacementList = NIL;
+		uint64 selectAnchorShardId = INVALID_SHARD_ID;
+		List *insertShardPlacementList = NULL;
+		List *intersectedPlacementList = NULL;
+		RangeTblEntry *rangeTableEntry = NULL;
+
+		/*
+		 * Replace the magic value in all baserestrictinfos. Note that
+		 * this has to be done on a copy, as the walker modifies in place.
+		 */
+		foreach(restrictionCell, copiedRestrictionContext->relationRestrictionList)
+		{
+			RelationRestriction *restriction = lfirst(restrictionCell);
+
+			restriction->relOptInfo->baserestrictinfo = (List *)
+														ReplaceHiddenParameter(
+				(Node *) restriction->relOptInfo->baserestrictinfo,
+				shardInterval);
+		}
+
+		/*
+		 * Use select planner to generate query for this specific
+		 * shard. We don't use the generated query, just rely on the
+		 * side-effect that all RTEs have been updated to point to the
+		 * relevant nodes.
+		 */
+		routerQuery = RouterSelectQuery(originalSubquery, subquery,
```

Comment:
Hm, I think what I dislike is the comment:
```
	 * shard. We don't use the generated query, just rely on the
	 * side-effect that all RTEs have been updated to point to the
	 * relevant nodes.
```
Because of it I hadn't realize that we don't generate the query anymore.  Note that it's not actually true that we don't use the returned query - we do, because it's just the passed in originalQuery.  Maybe we should rather make that routine return true/false?